src/disco/store/fd_pending_slots.c

```diff
@@ -73,127 +85,50 @@ fd_pending_slots_delete( void * pending_slots ) {
-    pending[slot & FD_PENDING_MASK] = when;
-
-  } else if ( slot < pending_slots->start ) {
-    /* Grow down */
-    if( (long)(pending_slots->end - slot) > (long)FD_PENDING_MAX )
-      FD_LOG_ERR(( "pending queue overrun: start=%lu, end=%lu, new slot=%lu", pending_slots->start, pending_slots->end, slot ));
-    pending[slot & FD_PENDING_MASK] = when;
-    for( ulong i = slot+1; i < pending_slots->start; i++ ) {
-      /* Zero fill */
-      pending[i & FD_PENDING_MASK] = 0;
-    }
-    pending_slots->start = slot;
-
-  } else if ( slot >= pending_slots->end ) {
-    /* Grow up */
-    if( (long)(slot - pending_slots->start) > (long)FD_PENDING_MAX )
-      FD_LOG_ERR(( "pending queue overrun: start=%lu, end=%lu, new slot=%lu", pending_slots->start, pending_slots->end, slot ));
-    pending[slot & FD_PENDING_MASK] = when;
-    for( ulong i = pending_slots->end; i < slot; i++ ) {
-      /* Zero fill */
-      pending[i & FD_PENDING_MASK] = 0;
+                      long when) {
+  fd_pending_slots_treap_ele_t *ele = fd_pending_slots_treap_ele_query( pending_slots->treap, slot, pending_slots->pool );
+  if ( ele ) {
+    if ( ele->time > when ) ele->time = when;
+  } else {
+    if ( fd_pending_slots_pool_free( pending_slots->pool )==0 ) {
+      FD_LOG_ERR(( "pending_slots (size=%d) is full during insertion", FD_BLOCK_MAX ));
     }
-    pending_slots->end = slot+1U;
 
-  } else {
-    /* Update in place */
-    long * p = &pending[slot & FD_PENDING_MASK];
-    if( 0 == *p || *p > when )
-      *p = when;
+    ele       = fd_pending_slots_pool_ele_acquire( pending_slots->pool );
+    ele->slot = slot;
+    ele->time = when;
+    ele->prio = fd_ulong_hash( ( pending_slots->total_insert++ ) ^ pending_slots->seed );
+    fd_pending_slots_treap_ele_insert( pending_slots->treap, ele, pending_slots->pool );
   }
-
-  fd_pending_slots_unlock( pending_slots );
 }
 
 long
 fd_pending_slots_get( fd_pending_slots_t * pending_slots,
                       ulong                slot ) {
-  if( pending_slots->start == pending_slots->end ) {
-    return LONG_MAX;
-  } else if( slot < pending_slots->start ) {
-    return LONG_MAX;
-  } else if( slot >= pending_slots->end ) {
-    return LONG_MAX;
-  } else {
-    return pending_slots->pending[ slot & FD_PENDING_MASK ];
-  }
+  fd_pending_slots_treap_ele_t *ele = fd_pending_slots_treap_ele_query( pending_slots->treap, slot, pending_slots->pool );
+  return ele ? ele->time : LONG_MAX;
 }
 
 void
 fd_pending_slots_set_lo_wmark( fd_pending_slots_t * pending_slots,
-                               ulong slot ) {
-  pending_slots->lo_wmark = slot;
+                               ulong lo_wmark ) {
+  fd_pending_slots_treap_fwd_iter_t iter;
+  for ( iter = fd_pending_slots_treap_fwd_iter_init( pending_slots->treap, pending_slots->pool );
+        !fd_pending_slots_treap_fwd_iter_done( iter ); ) {
+    fd_pending_slots_treap_ele_t * ele = fd_pending_slots_treap_fwd_iter_ele( iter, pending_slots->pool );
```

Comment:
rename to prev

Reply:
also no space after `for`, similar to `if`. see examples in kevin's code like here: https://github.com/firedancer-io/firedancer/blob/main/src/tango/mcache/fd_mcache.c#L65